waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,653 @@
+        (swap! service-id->failed-instances-transient-store dissoc service-id)
+        (scheduler/remove-killed-instances-for-service! service-id)
+        delete-result)
+      (catch [:status 404] _
+        (log/warn "service does not exist:" service-id)
+        {:result :no-such-service-exists
+         :message "Kubernetes reports service does not exist"})
+      (catch Throwable e
+        (log/warn "internal error while deleting service"
+                  {:service-id service-id})
+        {:result :error
+         :message "Internal error while deleting service"})))
+
+  (scale-app [this service-id scale-to-instances _]
+    (ss/try+
+      (if-let [service (service-id->service this service-id)]
+        (do
+          (scale-service-up-to this service scale-to-instances)
+          {:success true
+           :status 200
+           :result :scaled
+           :message (str "Scaled to " scale-to-instances)})
+        (do
+          (log/error "cannot scale missing service" service-id)
+          {:success false
+           :status 404
+           :result :no-such-service-exists
+           :message "Failed to scale missing service"}))
+      (catch [:status 409] _
+        {:success false
+         :status 409
+         :result :conflict
+         :message "Scaling failed due to repeated patch conflicts"})
+      (catch Throwable e
+        (log/error e "Error while scaling waiter service" service-id)
+        {:success false
+         :status 500
+         :result :failed
+         :message "Error while scaling waiter service"})))
+
+  (retrieve-directory-content [_ service-id instance-id _ relative-directory]
+    ;; TODO (#357) - get access to the working directory contents in the pod
+    [])
+
+  (service-id->state [_ service-id]
+    {:failed-instances (vals (get @service-id->failed-instances-transient-store service-id))
+     :killed-instances (scheduler/service-id->killed-instances service-id)})
+
+  (state [_]
+    {:service-id->failed-instances @service-id->failed-instances-transient-store}))
+
+(defn default-replicaset-builder
+  "Factory function which creates a Kubernetes ReplicaSet spec for the given Waiter Service."
+  [{:keys [orchestrator-name pod-base-port replicaset-api-version
+           service-id->password-fn] :as scheduler}
+   service-id
+   {:strs [backend-proto cmd cpus grace-period-secs health-check-interval-secs
+           health-check-max-consecutive-failures mem min-instances ports
+           run-as-user] :as service-description}
+   {:keys [container-image-spec] :as context}]
+  (let [home-path (str "/home/" run-as-user)
+        base-env (scheduler/environment service-id service-description
+                                        service-id->password-fn home-path)
+        ;; Randomize $PORT0 value to ensure clients can't hardcode it.
+        ;; Helps maintain compatibility with Marathon, where port assignment is dynamic.
+        port0 (-> (rand-int 100) (* 10) (+ pod-base-port))
```

Comment:
It would be nice to have a consistent `ReplicaSet` definition for a given `service-id`.  The "randomness" could be seeded.

Reply:
`port0` is now generated based on the `hash` of the `service-id` string:

```clojure
port0 (-> service-id hash (mod 100) (* 10) (+ pod-base-port))
```

This makes the port number deterministic for a given service description, but still practically unpredictable.